### PR TITLE
[ENG-1999] change pigeon

### PR DESCRIPTION
**Pull request by Johnetordoff**
## Purpose

So in order to combine all the different pigeon script and bundle it all up in a single importable package, I found there were a few hurdles in achieving this and I will enumerate them before explaining the changes I've made.

1. *File hierarchy problems*: As a bunch of different one-off scripts osf-pigeon could make directories and clean-up after them without any fuss, but when working together this process breaks down and directories conflict or reference a non-existent hierarchy 

2. *General cohesiveness*: While each script independently was working, none of the scripts had standardized parameters, so many required a local file path, or file names that could not be obtained as parameters. So I took some initiative in recomposing these scripts in an order that makes sense.

3. *Inconsistent mocking*: This had to mock Datacite, OSF api. waterbutler as well as a whole series of temporary files, this consolidated all those into an explicit test suite.

## Changes

- There's now a `pigeon` function that runs all these scripts together neatly.
- all file operation happen on `temp_dir` which is created by the pigeon function.
- Consuming logs, wikis and registration json is now all done by one function.
- Metadata modify method deleted, now the metadata is added on upload.

## QA Notes

⚠️ If you test using staging registrations make sure they have DOIs!

  - Does this change require a data migration? If so, what data will we migrate?
  - What is the level of risk?
    - Any permissions code touched?
    - Is this an additive or subtractive change, other?
  - How can QA verify? (Through UI, API, AdminApp or AdminAdminApp?)
    - If verifying through API, what's the new version? Please include the endpoints in PR notes or Dev docs.
  - What features or workflows might this change impact?
  - How will this impact performance?
-->

## Documentation

New readme.md !

## Side Effects

None that I know of.

## Ticket

https://openscience.atlassian.net/browse/ENG-1999

### CODE REVIEW

File: osf_pigeon/pigeon.py [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+                    'secret': ia_secret_key
+                }
+            }
+        )
+        ia_item = session.get_item(guid)
+        metadata = get_metadata(temp_dir, 'registration.json')
+
+        ia_item.upload(
+            {'bag.zip': zip_data},
+            metadata=metadata,
```
Comment (corbinSanders) [inline osf_pigeon/pigeon.py:161]: I appreciate the desire to consolidate the metadata upload into one function. However I am unsure if the IA library will update the metadata if the item already exists.

File: osf_pigeon/pigeon.py [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+            message='Too many requests, sleeping.',
+            period_remaining=sleep_period or int(resp.headers.get('Retry-After') or 0)
+        )  # This will be caught by @sleep_and_retry and retried
+    resp.raise_for_status()
+
+    return resp
+
+
+@sleep_and_retry
+def put_with_retry(
```
Comment (corbinSanders) [inline osf_pigeon/pigeon.py]: I don't see anywhere we need to do a put.

**Review by corbinSanders — CHANGES_REQUESTED** [review]
Looks good! Thank you for all the refactoring, John! I just have a few questions.

**Review by corbinSanders — APPROVED** [review]
LGTM! Thanks again!

File: osf_pigeon/__main__.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+if __name__ == '__main__':
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        '-g',
+        '--guid',
+        help='This is the GUID of the target node on the OSF',
+        required=True
+    )
+    parser.add_argument(
+        '-dp',
```
Comment (brianjgeiger) [inline osf_pigeon/__main__.py]: This should be either a single letter or should have a double-dash. `-p` should be fine.

File: osf_pigeon/__main__.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        '--datacite_password',
+        help='This is the password for using datacite\'s api',
+        required=False
+    )
+    parser.add_argument(
+        '-du',
```
Comment (brianjgeiger) [inline osf_pigeon/__main__.py]: This should be either a single letter or should have a double-dash. `-u` should be fine

File: osf_pigeon/__main__.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+        '--datacite_username',
+        help='This is the username for using datacite\'s api',
+        required=False
+    )
+    parser.add_argument(
+        '-iaak',
```
Comment (brianjgeiger) [inline osf_pigeon/__main__.py]: This should be either a single letter or should have a double-dash. `-a` should be fine.

File: osf_pigeon/__main__.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        '--ia_access_key',
+        help='This is the access key for using Internet Archive\'s api',
+        required=False
+    )
+    parser.add_argument(
+        '-iask',
```
Comment (brianjgeiger) [inline osf_pigeon/__main__.py]: This should be either a single letter or should have a double-dash. `-s` should be fine.

File: osf_pigeon/pigeon.py [hunk 7]
```diff
@@ -0,0 +1,243 @@
...
+            temp_dir,
+            guid,
+            datacite_username=datacite_username,
+            datacite_password=datacite_password,
+            datacite_prefix=datacite_prefix
+        )
+
+        zip_data = create_zip_data(temp_dir)
+
+        assert isinstance(ia_access_key, str),\
```
Comment (brianjgeiger) [inline osf_pigeon/pigeon.py]: Could these assertions come before we grab all of the data from OSF?

File: osf_pigeon/pigeon.py [hunk 8]
```diff
@@ -0,0 +1,243 @@
...
+        url,
+        retry_on: Tuple[int] = (),
+        sleep_period: int = None,
+        headers: Dict = None) -> requests.Response:
+
+    if not headers:
+        headers = {}
+
+    if not settings.OSF_THROTTLE_ENABLED:
+        assert settings.OSF_BEARER_TOKEN, 'must have OSF_BEARER_TOKEN set to disable throttle'
```
Comment (brianjgeiger) [inline osf_pigeon/pigeon.py]: Technically this will just make the throttling less likely. You could still be throttled if you're using a token.

**Review by brianjgeiger — CHANGES_REQUESTED** [review]
A few tiny things, and a couple of comments. Otherwise looks great.

**Review by brianjgeiger — APPROVED** [review]

**Pull request merged**